docs/enhancements/automatic-agent-classification-labels.md

```diff
@@ -0,0 +1,101 @@
+
+## Design Details [optional]
+
+We will create a new AgentClassification namespace-scoped CRD that defines the desired label key-value and a list of expressions to be evaluated according to the Agent’s inventory.  Each expression is generic such that it can refer to any properties in the inventory.
+
+### AgentClassification CRD Definition
+
+```
+// AgentClassificationSpec defines the desired state of AgentClassification
+type AgentClassificationSpec struct {
+	// LabelKey specifies the label key to apply to matched Agents
+	//
+	// +immutable
+	LabelKey string `json:"labelKey"`
+
+	// LabelValue specifies the label value to apply to matched Agents
+	//
+	// +immutable
+	LabelValue string `json:"labelValue"`
+
+	// ExpressionDefinitions define the conditions upon which the specified label is applied
+	ExpressionDefinitions []ExpressionDefinition `json:"expressionDefinitions"`
+}
+
+// ExpressionDefinition defines an expression to be evaluated
+type ExpressionDefinition struct {
+	// Expression is the expression that will be evaluated, in govaluate format
+	Expression string `json:"expression"`
+
+	// Parameters maps parameters from the above expression to paths in the Agent inventory
+	Parameters map[string]string `json:"parameters"`
+}
+```
+
+We will base the expression definition on the [govaluate](https://github.com/Knetic/govaluate) library, which supports a language for expressions with parameters.  Along with each expression (e.g., “cpu > 8”), the user will define a list of variables and map them to inventory items (e.g., “cpu”: “Cpu.Count”).  The controller will need to parse the inventory path and translate it into the corresponding value.  In addition, the inventory contains arrays for disks and network interfaces - the user will pass "Disks(N).SizeBytes" for at least N disks having the specified capacity, or "Disks().SizeBytes" for all disks.  The controller will need to parse this and create additional govaluate expressions for each disk, and then check for at least N returning true.
+
+### Operator changes
+
+When an Agent's inventory is updated, the Agent controller will now need to list all AgentClassification CRs in the namespace and update the Agent's labels accordingly.
+
+A new controller will watch AgentClassification CRs.  Upon creation/update, it will list all Agents in the namespace and apply the relevant labels.  It will also ensure that the specified label key is unique in that namespace.  Upon deletion, it will delete the label from all Agents.
+
+### Risks and Mitigations
+
+With a large number of Agents and AgentClassifications, the scale can be adversely affected.  To mitigate this, we scope the effects of an AgentClassifications to its namespace and evaluate the expressions only if the Agent's inventory was updated or an AgentClassification Spec was updated.  Both should not happen often.
```

Comment:
We keep adding logic and calculations around inventory changes.
The inventory gets updated on every inventory response from the agent (some optimization exists but doesn't seem to work)

Reply:
Yes, definitely.  I recently opened [MGMT-9343](https://issues.redhat.com/browse/MGMT-9343) to track this.